components/engine_rocks/src/write_batch.rs

```diff
@@ -7,134 +7,211 @@ use rocksdb::{Writable, WriteBatch as RawWriteBatch, DB};
     }
 
-    fn write_batch_with_cap(&self, cap: usize) -> RocksWriteBatch {
-        RocksWriteBatch::with_capacity(self, cap)
+    fn write_batch_with_cap(&self, cap: usize) -> RocksWriteBatchVec {
+        RocksWriteBatchVec::with_capacity(self, cap)
     }
 }
 
-pub struct RocksWriteBatch {
+/// `RocksWriteBatchVec` is for method `MultiBatchWrite` of RocksDB, which splits a large WriteBatch
+/// into many smaller ones and then any thread could help to deal with these small WriteBatch when it
+/// is calling `MultiBatchCommit` and wait the front writer to finish writing. `MultiBatchWrite` will
+/// perform much better than traditional `pipelined_write` when TiKV writes very large data into RocksDB.
+/// We will remove this feature when `unordered_write` of RocksDB becomes more stable and becomes compatible
+/// with Titan.
+pub struct RocksWriteBatchVec {
     db: Arc<DB>,
-    wb: RawWriteBatch,
+    wbs: Vec<RawWriteBatch>,
+    save_points: Vec<usize>,
+    index: usize,
+    cur_batch_size: usize,
+    batch_size_limit: usize,
+    support_write_batch_vec: bool,
 }
 
-impl RocksWriteBatch {
-    pub fn new(db: Arc<DB>) -> RocksWriteBatch {
-        let wb = RawWriteBatch::new();
-        RocksWriteBatch { db, wb }
-    }
-
-    pub fn with_capacity(engine: &RocksEngine, cap: usize) -> RocksWriteBatch {
+impl RocksWriteBatchVec {
+    pub fn new(db: Arc<DB>, batch_size_limit: usize, cap: usize) -> RocksWriteBatchVec {
         let wb = RawWriteBatch::with_capacity(cap);
-        RocksWriteBatch {
-            db: engine.as_inner().clone(),
-            wb,
+        RocksWriteBatchVec {
+            db: db.clone(),
+            wbs: vec![wb],
+            save_points: vec![],
+            index: 0,
+            cur_batch_size: 0,
+            batch_size_limit,
+            support_write_batch_vec: db.get_db_options().is_enable_multi_batch_write(),
         }
     }
 
-    pub fn as_inner(&self) -> &RawWriteBatch {
-        &self.wb
+    pub fn with_capacity(engine: &RocksEngine, cap: usize) -> RocksWriteBatchVec {
+        Self::new(engine.as_inner().clone(), WRITE_BATCH_LIMIT, cap)
+    }
+
+    pub fn as_inner(&self) -> &[RawWriteBatch] {
+        &self.wbs[0..=self.index]
     }
 
     pub fn as_raw(&self) -> &RawWriteBatch {
-        &self.wb
+        &self.wbs[0]
     }
 
     pub fn get_db(&self) -> &DB {
         self.db.as_ref()
     }
+
+    /// `check_switch_batch` will split a large WriteBatch into many smaller ones. This is to avoid
+    /// a large WriteBatch blocking write_thread too long.
+    fn check_switch_batch(&mut self) {
+        if !self.support_write_batch_vec {
+            return;
+        }
+        if self.batch_size_limit > 0 && self.cur_batch_size >= self.batch_size_limit {
+            self.index += 1;
+            self.cur_batch_size = 0;
+            if self.index >= self.wbs.len() {
+                self.wbs.push(RawWriteBatch::default());
+            }
+        }
```

Comment:
So that `count()` can always return `self.cur_batch_size + self.index * self.batch_size_limit`.
```suggestion
        if self.support_write_batch_vec {
            if self.batch_size_limit > 0 && self.cur_batch_size >= self.batch_size_limit {
                self.index += 1;
                self.cur_batch_size = 0;
                if self.index >= self.wbs.len() {
                    self.wbs.push(RawWriteBatch::default());
                }
            }
        }
```